### DISCUSSION TITLE
Re-activating a button

### DISCUSSION THREAD

**Discussion by KillTheMule** [fltk-rs/fltk-rs#996] [Q&A] [answered]
I have a button in my app that spawns a thread on callback. I want that button deactivated (both in fact and visually) during the runtime of the thread, and reactivated (again, also visually) after it's done. I'm currently doing it like this:

```rust
  render_but.set_callback({
    // clone some stuff for the closure
    move |button| {
      button.deactivate();
      button.set_color(Color::from_u32(GREY));
      button.set_label_color(Color::from_u32(BLACK));

   // get some values from text displays of the app

      std::thread::spawn({
        let mut button = button.clone();
        move || {
          some_long_running_stuff();

          button.set_label_color(Color::from_u32(YELLOW));
          button.set_color(Color::from_u32(BLUE));
          button.activate();
        }

      });
    }
  });
```

This kind of works, in that the button becomes grey when it's pressed, and it does not become blue again before the thread is done. *However*, it only becomes blue after moving the mouse (possibly other things might trigger it as well). If I don't move the mouse, I never get to see a blue button, even if the thread is long done. This happens on both linux and windows. I'm using fltk-1.2.16 as of now, and fltk-theme-0.4.2.

I tried various combinations of `redraw`ing either the app, the window, or the widget to no avail.

Any hint welcome!

**Comment by MoAlyousef** [comment, marked as answer]
You can add `app::awake();` after calling `some_long_running_stuff();`:
```rust
        move || {
          some_long_running_stuff();
          app::awake();
          button.set_label_color(Color::from_u32(YELLOW));
          button.set_color(Color::from_u32(BLUE));
          button.activate();
        }
```

Pre 1.0, all widget methods called awake, but this caused some busy loops in threads in some situations.

**Comment by KillTheMule** [comment]
That worked out nicely, thanks a bunch!
